File: ts/src/claudecode/process.ts
import { spawn as nodeSpawn } from "node:child_process";
import type { ClaudeRunnerConfig, SpawnFn } from "./options.js";

/** Log the command about to be executed. */
export function logCmd(
  config: ClaudeRunnerConfig,
  binary: string,
  args: string[],
  cwd?: string,
): void {
  if (!config.logger) return;

  const quoted = [binary, ...args].map((a) => {
    if (/[\s"'\\]/.test(a)) {
      return `'${a.replace(/'/g, "'\\''")}'`;
    }
    return a;
  });
  config.logger.debug("executing CLI command", {
    cmd: quoted.join(" "),
    dir: cwd ?? "",
  });
}

/** Resolve the spawn function, using the injected one or the default. */
export function resolveSpawn(config: ClaudeRunnerConfig): {
  spawn: SpawnFn;
  binary: string;
} {
  const binary = config.binary ?? "claude";

  if (config.spawn) {
    return { spawn: config.spawn, binary };
  }

  const defaultSpawn: SpawnFn = (command, args, options) => {
    return nodeSpawn(command, args, {
      cwd: options.cwd,
      env: options.env,
      signal: options.signal,
      stdio: ["ignore", "pipe", "pipe"],
    });
  };

  return { spawn: defaultSpawn, binary };
}

/**
 * Build a human-readable error string from stderr and unparseable stdout lines.
 */
export function collectErrorDetail(
  stderr: string,
  stdoutErrors: string[],
): string {
  const parts: string[] = [];
  const trimmed = stderr.trim();
  if (trimmed) {
    parts.push(trimmed);
  }
  if (stdoutErrors.length > 0) {
    parts.push(stdoutErrors.join("\n"));
  }
  if (parts.length === 0) {
    return "unknown error (no output from CLI)";
  }
  return parts.join("\n");
}

File: python/src/agentrunner/claudecode/process.py
"""Process helpers: logging, error collection, spawn resolution."""

from __future__ import annotations

import asyncio
import shlex
from typing import Protocol

from .options import Logger


class SpawnFn(Protocol):
    """Function that spawns an async subprocess.

    Internal protocol used for dependency injection in tests.
    """

    async def __call__(
        self,
        program: str,
        *args: str,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
    ) -> asyncio.subprocess.Process: ...


def log_cmd(
    logger: Logger | None,
    binary: str,
    args: list[str],
    cwd: str | None = None,
) -> None:
    """Log the command about to be executed."""
    if not logger:
        return

    cmd = " ".join(shlex.quote(a) for a in [binary, *args])
    logger.debug("executing CLI command", extra={"cmd": cmd, "dir": cwd or ""})


async def default_spawn(
    program: str,
    *args: str,
    cwd: str | None = None,
    env: dict[str, str] | None = None,
) -> asyncio.subprocess.Process:
    return await asyncio.create_subprocess_exec(
        program,
        *args,
        cwd=cwd,
        env=env,
        stdin=asyncio.subprocess.DEVNULL,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )


def collect_error_detail(stderr: str, stdout_errors: list[str]) -> str:
    """Build a human-readable error string from stderr and unparseable stdout lines."""
    parts: list[str] = []
    trimmed = stderr.strip()
    if trimmed:
        parts.append(trimmed)
    if stdout_errors:
        parts.append("\n".join(stdout_errors))
    if not parts:
        return "unknown error (no output from CLI)"
    return "\n".join(parts)
